### PR TITLE
Qsv to use tiled aligner rather than BLAT

### PR DESCRIPTION
**Pull request by holmeso**
# Description

This PR covers the change in `qsv` from using BLAT to using the tiled aligner (TA).
There are a number of places within qsv where a call is made to BLAT. These have been replaced with a call to the TA.

There are also a number of enhancements to the TA itself after extensive testing against a cohort of 50 melanoma donors.
This was to reduce the number of unique calls to both BLAT and TA.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Additional unit tests have been provided to cover any new methods.
Process has been run against 50 melanoma donors repeatedly.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: q3tiledaligner/src/au/edu/qimr/tiledaligner/util/TARecordUtil.java [hunk 1]
```diff
@@ -483,6 +132,97 @@ public static TLongList getLongList(long ... list) {
...
+				if (null != blatties && blatties.length > 0) {
+					blats.add(blatties);
+				}
+			}
+			return blats;
+		}
+		return Collections.emptyList();
+	}
+	
 	public static String[] blatRecordFromSplit(IntLongPair split, String name, int seqLength, PositionChrPositionMap headerMap) {
```
Comment (delocalizer) [inline q3tiledaligner/src/au/edu/qimr/tiledaligner/util/TARecordUtil.java:226]: A bit pedantic perhaps, but the name of this implies it returns BLATRecord - is there a reason we're returning the raw String array and not the constructed BLATRecord?

File: q3tiledaligner/src/au/edu/qimr/tiledaligner/util/TiledAlignerUtil.java [hunk 2]
```diff
@@ -1460,20 +1632,657 @@ public static boolean doesSequenceHaveMostlySingleBaseRepeats(String sequence) {
...
+		 * SW wasn't able to return records within the filtering criteria
+		 * Lets see if we can manually create one based on the tile count
+		 */
+		if ((length - mismatchCount) > MINIMUM_BLAT_RECORD_SCORE) {
+			
+			logger.debug("SW couldn't put this down whilst passing the filters - will attempt to create based on tile counts alone. l: " + l + ", length: " + length + ", mismatchCount: " + mismatchCount);
+			
+			ChrPosition cp = headerMap.getChrPositionFromLongPosition(l);
+			boolean forwardStrand = "F".equals(((ChrPositionName)cp).getName());
+//			int buffer = 10;
```
Comment (delocalizer) [inline q3tiledaligner/src/au/edu/qimr/tiledaligner/util/TiledAlignerUtil.java:2248]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/TARecordTest.java [hunk 3]
```diff
@@ -181,9 +181,8 @@ public void getCountWithLotsOfCommonTiles() {
 		for (Entry<Integer, TLongList> entry : map2.entrySet()) {
 			System.out.println("entry getKey: " + Arrays.toString(NumberUtils.splitIntInto2(entry.getKey())) + ", list size: " + entry.getValue().size());
 		}
-		assertEquals(2, map2.size());
-		assertEquals(true, map2.containsKey(NumberUtils.pack2IntsInto1(18, 0)));
-		assertEquals(true, map2.containsKey(NumberUtils.pack2IntsInto1(20, 0)));
+		assertEquals(1, map2.size());
+		assertEquals(true, map2.containsKey(NumberUtils.pack2IntsInto1(67, 0)));
 //		assertArrayEquals(new long[] {1240342426l}, map2.get(NumberUtils.pack2IntsInto1(18, 0)).toArray());
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/TARecordTest.java:186]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 4]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
-		assertEquals("281	286	2	0	0	1	10	2	13	+	splitcon_chr19_21832833_chr19_21916389_4_true_1599654284052_407169	358	12	307	chr19	12345	21916349	21916648	3	32,178,76	12,46,232	21916349,21916383,21916572", results.get(1).toString());
-		assertEquals("227	233	2	0	0	1	55	3	60	-	splitcon_chr19_21832833_chr19_21916389_4_true_1599654284052_407169	358	234	55	chr19	12345	21832520	21832813	4	66,82,41,44	58,170,256,302	21832520,21832635,21832723,21832769", results.get(2).toString());
+		assertEquals("66	68	0	0	0	1	4	1	32	+	splitcon_chr8_39360286_chr8_39360413__true_1603194926354_561727	403	328	399	chr8	12345	39360216	39360316	2	34,34	328,366	39360216,39360282", results.get(0).toString());
+		assertEquals("330	336	2	0	0	1	14	3	26	+	splitcon_chr8_39360286_chr8_39360413__true_1603194926354_561727	403	50	399	chr8	12345	39359916	39360278	4	204,70,28,34	50,254,328,366	39359916,39360142,39360216,39360244", results.get(1).toString());
+		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplits(splits, name, seq.length(), pcpm);
+		assertEquals(1, blatRecs.size());
+		
+//		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:1695]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 5]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
+		
+		System.out.println("Number of records in results after looking in potentialSplits list: " + results.size());
+		
+		assertEquals(1, results.size());
+		assertEquals(1, results.stream().filter(br -> br.getScore() >= passingScore).count());
+		assertEquals("181	182	0	0	0	0	0	1	308	+	splitcon_chr5_178939187_chr5_178939496__true_1603195044083_189550_clip	195	13	194	chr5	12345	178939063	178939553	2	110,72	13,123	178939063,178939481", results.get(0).toString());
 		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplits(splits, name, seq.length(), pcpm);
 		assertEquals(1, blatRecs.size());
 		
+//		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:1750]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 6]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
+		
+		System.out.println("Number of records in results after looking in potentialSplits list: " + results.size());
+		
+		assertEquals(1, results.size());
+		assertEquals(1, results.stream().filter(br -> br.getScore() >= passingScore).count());
+		assertEquals("168	169	0	0	0	0	0	1	301	+	splitcon_chr14_64866810_chr14_64867112__true_1603769470012_401600_clip	169	0	168	chr14	12345	64866772	64867242	2	105,64	0,105	64866772,64867178", results.get(0).toString());
+		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplits(splits, name, seq.length(), pcpm);
+		assertEquals(1, blatRecs.size());
+		
+//		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:1806]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 7]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
+		assertEquals(0, results.size());
+		assertEquals(0, results.stream().filter(br -> br.getScore() >= passingScore).count());
+//		assertEquals("168	169	0	0	0	0	0	1	301	+	splitcon_chr14_64866810_chr14_64867112__true_1603769470012_401600_clip	169	0	168	chr14	12345	64866772	64867242	2	105,64	0,105	64866772,64867178", results.get(0).toString());
+		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplitsNew(splits, name, seq.length(), pcpm);
+		assertEquals(1, blatRecs.size());
+		assertEquals(2, blatRecs.get(0).length);
+		assertEquals("21	21	0	0	0	0	0	0	0	+	splitcon_chr17_80666697_chr17_80666858__true_1603769669765_514144	312	3	24	chr17	12345	36324761	36324782	1	21	3	36324761", blatRecs.get(0)[0].toString());
+		assertEquals("258	261	2	0	0	0	0	1	160	+	splitcon_chr17_80666697_chr17_80666858__true_1603769669765_514144	312	45	305	chr17	12345	80666575	80666996	2	132,129	45,177	80666575,80666867", blatRecs.get(0)[1].toString());
+		
+//		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:1867]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 8]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
+		assertEquals(1, results.size());
+		assertEquals(0, results.stream().filter(br -> br.getScore() >= passingScore).count());
+//		assertEquals("168	169	0	0	0	0	0	1	301	+	splitcon_chr14_64866810_chr14_64867112__true_1603769470012_401600_clip	169	0	168	chr14	12345	64866772	64867242	2	105,64	0,105	64866772,64867178", results.get(0).toString());
+		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplitsNew(splits, name, seq.length(), pcpm);
+		assertEquals(1, blatRecs.size());
+		assertEquals(2, blatRecs.get(0).length);
+		assertEquals("25	25	0	0	0	0	0	0	0	+	splitcon_chr7_25345087_chr7_25345211__false_1604284945574_644092	371	33	58	chr7	12345	148563703	148563728	1	25	33	148563703", blatRecs.get(0)[0].toString());
+		assertEquals("242	245	0	0	0	1	2	2	123	+	splitcon_chr7_25345087_chr7_25345211__false_1604284945574_644092	371	124	370	chr7	12345	25344983	25345351	3	104,46,95	124,228,276	25344983,25345210,25345256", blatRecs.get(0)[1].toString());
+		
 //		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:1978]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 9]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
+		
+		assertEquals(1, results.size());
+		assertEquals(1, results.stream().filter(br -> br.getScore() >= passingScore).count());
+		assertEquals("252	255	2	0	0	0	0	1	159	+	splitcon_chr6_153932501_chr6_153932720_1_true_1606270441498_475180	261	6	260	chr6	12345	153932406	153932820	2	144,111	6,150	153932406,153932709", results.get(0).toString());
+		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplitsNew(splits, name, seq.length(), pcpm);
+		assertEquals(1, blatRecs.size());
+		assertEquals(1, blatRecs.get(0).length);
+		assertEquals("252	255	2	0	0	0	0	1	159	+	splitcon_chr6_153932501_chr6_153932720_1_true_1606270441498_475180	261	6	260	chr6	12345	153932406	153932820	2	144,111	6,150	153932406,153932709", blatRecs.get(0)[0].toString());
+		
+//		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:2037]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 10]
```diff
@@ -1682,26 +1679,483 @@ public void realLifeSplits27() {
...
+		
+		assertEquals(1, results.size());
+		assertEquals(1, results.stream().filter(br -> br.getScore() >= passingScore).count());
+		assertEquals("200	204	2	0	0	0	0	2	74	+	splitcon_chr6_153932501_chr6_153932720_1_true_1606270441498_475180	204	0	203	chr3	12345	116488150	116488428	3	70,73,61	0,70,143	116488150,116488296,116488367", results.get(0).toString());
+		List<BLATRecord[]> blatRecs = TARecordUtil.blatRecordsFromSplitsNew(splits, name, seq.length(), pcpm);
+		assertEquals(1, blatRecs.size());
+		assertEquals(1, blatRecs.get(0).length);
+		assertEquals("200	204	2	0	0	0	0	2	74	+	splitcon_chr6_153932501_chr6_153932720_1_true_1606270441498_475180	204	0	203	chr3	12345	116488150	116488428	3	70,73,61	0,70,143	116488150,116488296,116488367", blatRecs.get(0)[0].toString());
+		
+//		for (BLATRecord[] brs : blatRecs) {
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:2095]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java [hunk 11]
```diff
@@ -1867,14 +2321,15 @@ public void realLifeNotASplit7() {
...
-//		}
-		assertEquals(2, blatRecs.get(0).length);
-		assertEquals("24	24	0	0	0	0	0	0	0	-	chr4_65841131_false_-	57	0	24	chr4	12345	65841100	65841124	1	24	33	65841100", blatRecs.get(0)[0].toString());
-		assertEquals("36	36	0	0	0	0	0	0	0	-	chr4_65841131_false_-	57	16	52	chr4	12345	65841104	65841140	1	36	5	65841104", blatRecs.get(0)[1].toString());
+		for (BLATRecord br : blatRecs.get(0)) {
+			System.out.println("blat record: " + br.toString());
+		}
+		assertEquals(1, blatRecs.get(0).length);
+		assertEquals("44	46	0	0	0	1	1	1	1	-	chr4_65841131_false_-	57	25	23	chr4	12345	65841077	65841124	2	22,24	10,33	65841077,65841100", blatRecs.get(0)[0].toString());
+//		assertEquals("24	24	0	0	0	0	0	0	0	-	chr4_65841131_false_-	57	0	24	chr4	12345	65841100	65841124	1	24	33	65841100", blatRecs.get(0)[0].toString());
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TARecordUtilTest.java:2331]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TiledAlignerUtilTest.java [hunk 12]
```diff
@@ -904,23 +954,100 @@ public void shortcut() {
...
+		assertEquals(1, shortcuts.size());
+		assertEquals(1176804192, shortcuts.keys()[0]);
+	}
+	
+	@Test
+	public void shortcutWithBuffer2() {
+		long[][] array = new long[3][];
+		
+		array = new long[146][];
+//		array[0] = new long[]{1319501472l};
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TiledAlignerUtilTest.java:1007]: ndc

File: q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TiledAlignerUtilTest.java [hunk 13]
```diff
@@ -904,23 +954,100 @@ public void shortcut() {
...
+		array[0] = new long[]{1176804192l};
+		array[67] = new long[]{90351255, 439695611, 777771917, 1094002331, 1167523024, 1176804877, 1228579535, 1354039040, 1561128813, 1856882536, 1870698140, 1890031905, 2113951585, 2617800701l, 2683547413l, 2742508068l, 2770451401l};
+		TLongIntMap shortcuts = TiledAlignerUtil.getShortCutPositionsForSmithwaterman(array, 500000);
+//		TLongList shortcuts = TiledAlignerUtil.getShortCutPositionsForSmithwaterman(array, 500000);
```
Comment (delocalizer) [inline q3tiledaligner/test/au/edu/qimr/tiledaligner/util/TiledAlignerUtilTest.java:985]: ndc

File: qcommon/src/org/qcmg/common/sw/SmithWatermanGotoh.java [hunk 14]
```diff
@@ -190,25 +192,34 @@ private void fillMatrix() {
...
-				verticalGaps[row][column] = 1;
-				horizontalGaps[row][column] = 1;
-			}
-		}		
+		for (int row = 0; row < rows; row++) {
+			Arrays.fill(verticalGaps[row], ONE);
+			Arrays.fill(horizontalGaps[row], ONE);
+		}
 	}
+//	private void initialize() {
```
Comment (delocalizer) [inline qcommon/src/org/qcmg/common/sw/SmithWatermanGotoh.java:203]: ndc

File: qcommon/src/org/qcmg/common/sw/SmithWatermanGotoh.java [hunk 15]
```diff
@@ -363,6 +374,70 @@ public static void main(String[] args) throws IOException {
...
+		
+		String r = "GGTCCACTCTACTCCAATATGATCTCATTTCAATCCTTACCTTAATTATACCTGCAAAGACTCTATTTCCAAATAAGGTCACATTCTGAGGTTCTCGACGAACATGAATTTTGGGTGGGCACTATTCAATCCACTATATAAATTTAACACACCATTTTAGTATTGACCATAACATGTTATAAAATTTCATATCATTGAATTATTTATACATATTATTTGTTAACACGAATCAGT";
+		String s = "GGTCCACTCTACTCCAATATGATCTCATTTCAATCCTTACCTTAATTATACCTGCAAAGACTGGTCCACTCTACTCCAATATGATCTCATTTCAATCCTTACCTTAATTATACCTGCAAAGACTGGTCCACTCTACTCCAATATGATCTCATTTCAATCCTTACCTTAATTATA";
+		nm = new SmithWatermanGotoh(r, s,    5, -4, 16, 4);
+		results = nm.traceback();
+		for (String s1 : results) {
+			System.out.println(s1);
+		}
+		nm = new SmithWatermanGotoh(r, s,    4, -14, 14, 1);
+//		nm = new SmithWatermanGotoh(r, s,   5, -4, 16, 4);
```
Comment (delocalizer) [inline qcommon/src/org/qcmg/common/sw/SmithWatermanGotoh.java:387]: ndc

File: qcommon/src/org/qcmg/common/util/NumberUtils.java [hunk 16]
```diff
@@ -289,13 +290,107 @@ public static int getPositionOfLongInArray(long [] array, long value) {
...
+		Arrays.sort(existingRanges, Comparator.comparingInt(arr -> arr[0]));
+		
+		/*
+		 * now loop through and add to list if there the gap between ranges or boundaries is large enough
+		 */
+		List<int[]> ranges = new ArrayList<>(4);
+		int lastEndPosition = 0;
+		for (int [] existingRange : existingRanges) {
+				
+//				int startPosition = NumberUtils.getShortFromLong(pair.getLong(), TILE_OFFSET);
```
Comment (delocalizer) [inline qcommon/src/org/qcmg/common/util/NumberUtils.java:308]: ndc

File: qsv/src/org/qcmg/qsv/QSVCluster.java [hunk 17]
```diff
@@ -592,26 +605,76 @@ public SoftClipCluster getPrimarySoftClipCluster() {
 	 * @return
 	 * @throws Exception
 	 */
-	public boolean findSplitReadContig(BLAT blat, QSVParameters p, QSVParameters n, String softclipDir, int consensusLength, boolean isQCMG, int minInsertSize, 
-			boolean singleSided, boolean isSplitRead, String referenceFile) throws Exception {
+//	public boolean findSplitReadContig(QSVParameters p, QSVParameters n, String softclipDir, int consensusLength, boolean isQCMG, int minInsertSize, boolean singleSided, boolean isSplitRead, String referenceFile) throws Exception {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/QSVCluster.java:608]: ndc

File: qsv/src/org/qcmg/qsv/QSVCluster.java [hunk 18]
```diff
@@ -631,6 +694,45 @@ public boolean findSplitReadContig(BLAT blat, QSVParameters p, QSVParameters n,
 		this.rescued = true;
 		return rescue;
 	}
+//	public boolean findSplitReadContig(BLAT blat, QSVParameters p, QSVParameters n, String softclipDir, int consensusLength, boolean isQCMG, int minInsertSize,
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/QSVCluster.java:697]: ndc

File: qsv/src/org/qcmg/qsv/QSVCluster.java [hunk 19]
```diff
@@ -724,19 +826,33 @@ public boolean isPotentialSplitRead() {
 	/*
 	 * For one sided clips, go back to bam to see if clips at the other breakpoint can be rescues
 	 */
-	private void rescueClipping(BLAT blat, QSVParameters p, QSVParameters n, String softclipDir, int consensusLength, int minInsertSize) throws Exception {
+//	private void rescueClipping(BLAT blat, QSVParameters p, QSVParameters n, String softclipDir, int consensusLength, int minInsertSize) throws Exception {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/QSVCluster.java:829]: ndc

File: qsv/src/org/qcmg/qsv/QSVCluster.java [hunk 20]
```diff
@@ -1162,15 +1278,27 @@ public boolean printRecord(boolean isSingleSided) {
 	 * @param minInsertSize
 	 * @throws Exception
 	 */
-	public void rescueClippping(BLAT blat, QSVParameters tumourParameters,
-			QSVParameters normalParameters, String softclipDir,
+//	public void rescueClippping(BLAT blat, QSVParameters tumourParameters,
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/QSVCluster.java:1281]: ndc

File: qsv/src/org/qcmg/qsv/QSVCluster.java [hunk 21]
```diff
@@ -1191,31 +1319,63 @@ public void rescueClippping(BLAT blat, QSVParameters tumourParameters,
...
+					this.normalSplitReadContig = new SplitReadContig(cache, normalParameters, softclipDir, leftReference, rightReference, 
 								getLeftBreakpoint(), getRightBreakpoint(), 
 								findExpectedPairClassifications(), getClipContigSequence(), getConfidenceLevel(), 
-								getOrientationCategory(), reference, normalParameters.getChromosomes(), hasSoftClipEvidence(), blatFile + ".normal.fa");
+								getOrientationCategory(), reference, normalParameters.getChromosomes(), hasSoftClipEvidence(), blatFile + ".normal.fa", log);
 				}						
 			}
 		}		
 	}
+//	public void createSplitReadContig(BLAT blat,
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/QSVCluster.java:1354]: ndc

File: qsv/src/org/qcmg/qsv/QSVPipeline.java [hunk 22]
```diff
@@ -479,16 +501,17 @@ private void findPairClusters() throws QSVException {
 	 */
 	private void findSoftClips() throws Exception {
 
-		BLAT blat = new BLAT(options.getBlatServer(), options.getBlatPort(), options.getBlatPath());
+//		BLAT blat = new BLAT(options.getBlatServer(), options.getBlatPort(), options.getBlatPath());
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/QSVPipeline.java:504]: ndc

File: qsv/src/org/qcmg/qsv/discordantpair/FindDiscordantPairClustersMT.java [hunk 23]
```diff
@@ -187,7 +187,7 @@ private void processByMatePairList() throws Exception {
 //			currentCluster.getClusterMatePairs().add(startPair);
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/discordantpair/FindDiscordantPairClustersMT.java:187]: ndc

File: qsv/src/org/qcmg/qsv/softclip/Breakpoint.java [hunk 24]
```diff
@@ -392,26 +395,56 @@ public static ConsensusRead calculate(char strand, int clipSize, int length, int
 		}
 	}
 
+//	public static String calculateConsensus(char strand, int length, boolean isClipConsensus, Set<Clip> tumourClips, Set<Clip> normalClips, boolean isGermline) {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/softclip/Breakpoint.java:398]: ndc

File: qsv/src/org/qcmg/qsv/softclip/Breakpoint.java [hunk 25]
```diff
@@ -568,29 +601,47 @@ public static String getClipString(Set<Clip> clips) {
 		return clips.stream().map(Clip::toString).collect(Collectors.joining(Constants.EMPTY_STRING));
 	}
 
-	public boolean findRescuedMateBreakpoint(BLAT blat, QSVParameters p, String softclipDir) throws Exception {
-		
-		String base = softclipDir + QSVUtil.getFileSeparator() + getName();
-		String fa = base + ".fa";
+//	public boolean findRescuedMateBreakpoint(BLAT blat, QSVParameters p, String softclipDir) throws Exception {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/softclip/Breakpoint.java:604]: ndc

File: qsv/src/org/qcmg/qsv/softclip/FindClipClustersMT.java [hunk 26]
```diff
@@ -1024,6 +1024,23 @@ public RescueRecord(AbstractQueue<List<QSVCluster>> queueIn2,
 			this.isSplitRead = isSplitRead;
 			this.reference = reference;
 		}
+//		public RescueRecord(AbstractQueue<List<QSVCluster>> queueIn2,
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/softclip/FindClipClustersMT.java:1027]: ndc

File: qsv/src/org/qcmg/qsv/softclip/FindClipClustersMT.java [hunk 27]
```diff
@@ -1060,25 +1077,39 @@ public RescueRecord(AbstractQueue<List<QSVCluster>> queueIn2,
...
+							logger.debug("rescue clipping called findSplitReadContig, getConfidenceLevel: " + cluster.getConfidenceLevel(true));
+							logger.debug("rescue clipping called findSplitReadContig, isPotentialSplitRead: " + cluster.isPotentialSplitRead());
 							clusters.add(cluster);
 						}
 
-						new File(blatFile + ".tumour.fa").delete();
-						new File(blatFile + ".tumour.psl").delete();
-						new File(blatFile + ".normal.fa").delete();
-						new File(blatFile + ".normal.psl").delete();
+//						new File(blatFile + ".tumour.fa").delete();
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/softclip/FindClipClustersMT.java:1109]: ndc

File: qsv/src/org/qcmg/qsv/softclip/SoftClipCluster.java [hunk 28]
```diff
@@ -417,7 +417,22 @@ public String getRightInfo() {
 		return this.name + TAB + rightBreakpoint + TAB + rightStrand + TAB + rightMateStrand + TAB + size; 
 	}
 
-	public void rescueClips(QSVParameters p, BLAT blat, File tumourFile, File normalFile, String softclipDir, int consensusLength, int chrBuffer, Integer minInsertSize) throws Exception {
+//	public void rescueClips(QSVParameters p, BLAT blat, File tumourFile, File normalFile, String softclipDir, int consensusLength, int chrBuffer, Integer minInsertSize) throws Exception {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/softclip/SoftClipCluster.java:420]: ndc

File: qsv/src/org/qcmg/qsv/softclip/SoftClipCluster.java [hunk 29]
```diff
@@ -489,6 +504,62 @@ public void findMaxRescueBreakpoint(QSVParameters p, BLAT blat, Map<Integer, Bre
 			} 
 		}
 	}	
+//	public void findMaxRescueBreakpoint(QSVParameters p, BLAT blat, Map<Integer, Breakpoint> leftMap,
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/softclip/SoftClipCluster.java:507]: ndc

File: qsv/src/org/qcmg/qsv/splitread/SplitReadContig.java [hunk 30]
```diff
@@ -342,31 +365,160 @@ private void assemble(List<UnmappedRead> splitReads, String clipContig, String s
 
 	}
 
-	public void findSplitRead() throws Exception {
-		if (consensus != null) {
+//	public void findSplitRead() throws Exception {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/splitread/SplitReadContig.java:368]: ndc

File: qsv/src/org/qcmg/qsv/splitread/SplitReadContig.java [hunk 31]
```diff
@@ -774,23 +980,52 @@ private void getBestBlocksFromBLAT(BLATRecord record) {
...
+	private void setSplitReadAlignments(BLATRecord r1, BLATRecord r2) {
+		/*
+		 * call the static method
+		 */
+		Pair<SplitReadAlignment, SplitReadAlignment> pair = setSplitReadAlignments(r1, r2, knownSV.getLeftBreakpoint().intValue(), knownSV.getRightBreakpoint().intValue(), confidenceLevel, length);
+		left = pair.getLeft();
+		right = pair.getRight();
+	}
+
+//	private void setSplitReadAlignments(BLATRecord r1, BLATRecord r2) {
```
Comment (delocalizer) [inline qsv/src/org/qcmg/qsv/splitread/SplitReadContig.java:1013]: ndc

File: qsv/test/org/qcmg/qsv/softclip/BreakpointTest.java [hunk 32]
```diff
@@ -469,32 +468,32 @@ public void compare() {
...
-		replay(blat);
-        QSVParameters p = createMock(QSVParameters.class);             
-        
-        assertTrue(breakpoint.findRescuedMateBreakpoint(blat, p, softClipDir));
-        assertEquals(QSVUtil.MINUS, breakpoint.getMateStrand());
-        // not so because we have now set isLeft to be true, rather than the default value which is false
-        assertEquals(89700252, breakpoint.getMateBreakpoint());
-        assertEquals("chr10", breakpoint.getMateReference());
-	}
+//	@Test
```
Comment (delocalizer) [inline qsv/test/org/qcmg/qsv/softclip/BreakpointTest.java:471]: ndc

File: qsv/test/org/qcmg/qsv/softclip/FindSoftClipClustersMTTest.java [hunk 33]
```diff
@@ -149,11 +151,12 @@ public void testClipClusterWorker() throws Exception {
...
         replay(blat);
 		
-		FindClipClustersMT worker = new FindClipClustersMT(tumor, normal, softClipDir.getAbsolutePath(), blat, tumorClusterRecords,  options, "analysisId", 200);
+//		FindClipClustersMT worker = new FindClipClustersMT(tumor, normal, softClipDir.getAbsolutePath(), blat, tumorClusterRecords,  options, "analysisId", 200);
+		FindClipClustersMT worker = new FindClipClustersMT(tumor, normal, softClipDir.getAbsolutePath(), tumorClusterRecords,  options, "analysisId", 200, new TIntObjectHashMap<int[]>());
         worker.execute();
         
-        assertEquals(0, worker.getExitStatus().intValue());
-        assertEquals(2, worker.getQSVRecordWriter().getSomaticCount().intValue());
+//        assertEquals(0, worker.getExitStatus().intValue());
```
Comment (delocalizer) [inline qsv/test/org/qcmg/qsv/softclip/FindSoftClipClustersMTTest.java:158]: ndc

File: qsv/test/org/qcmg/qsv/softclip/SoftClipClusterTest.java [hunk 34]
```diff
@@ -232,38 +232,38 @@ public void testGetClipCount() throws Exception {
...
-						.getAbsolutePath(),
-						testFolder.newFile("chr10_89712341_true_+.psl")
-								.getAbsolutePath())).andReturn(expected);
-		replay(blat);
-
-		clip.rescueClips(tumor, blat, new File(tumourFile),
-				new File(normalFile), testFolder.getRoot().toString(), 20, 200,
-				50);
-	}
+//	@Test
```
Comment (delocalizer) [inline qsv/test/org/qcmg/qsv/softclip/SoftClipClusterTest.java:235]: ndc

File: qsv/test/org/qcmg/qsv/splitread/SplitReadContigTest.java [hunk 35]
```diff
@@ -429,12 +602,25 @@ public void testTranslocationOrientation4() throws Exception {
 			"TGAGTTCATTGATCACCTTCTGCTTCTCCACCAGCATCTTGTCCTTCTCGGAGTCCAGCCCATCCAGCTCAGAAGGAACAGAGCCCCCATTCAGGCTGCTAAAAAGCTCAGCGCAAAGCGTGCGGATCTGCAGTCCACCTTCTCTGGAGGACGAATTCCAAAGAAGTTTGCCCGCAGAGGCACCAGCCTCAAAGAACGGCTGTG");		
 	}
 
+//	private void testOrientationCategory(String leftReference, String rightReference,
```
Comment (delocalizer) [inline qsv/test/org/qcmg/qsv/splitread/SplitReadContigTest.java:605]: ndc

**Review by delocalizer — APPROVED** [review]
This is terrific. I feel privileged to be present for this one :)

**Comment by delocalizer** [comment]
@ChristinaXu2017 I would have liked ollie's responses to my questions & comments before merging this.

**Comment by ChristinaXu2017** [comment]
sorry, I just want to merge these changes into my branch, so I can update the regression tests. it is better to leave comments here but not approve If you are waiting for the response.

**Pull request merged**